server/src/server.ts

```diff
@@ -383,11 +385,49 @@ function onMessage(msg: m.Message) {
         // error: code and message set in case an exception happens during the definition request.
       };
       send(definitionResponse);
+    } else if (msg.method === p.RenameRequest.method) {
+      // https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_rename
+      let params = msg.params as p.RenameParams;
+      let filePath = fileURLToPath(params.textDocument.uri);
+      let locations: Location[] | null = utils.runAnalysisAfterSanityCheck(
+        filePath,
+        [
+          "references",
+          filePath,
+          params.position.line,
+          params.position.character,
+        ]
+      );
+
+      let result: WorkspaceEdit | null;
+      if (locations === null) {
+        result = null;
+      } else {
+        let changes: { [uri: string]: TextEdit[] } = {};
+        locations.forEach(({ uri, range }) => {
+          let textEdit: TextEdit = {range, newText: params.newName};
+          if (uri in changes) {
+            changes[uri].push(textEdit);
+          } else {
+            changes[uri] = [textEdit]
+          }
+        });
+
+        result = {changes};
+      }
+
+      let renameResponse: m.ResponseMessage = {
+        jsonrpc: c.jsonrpcVersion,
+        id: msg.id,
+        result,
+      };
+
+      send(renameResponse);
     } else if (msg.method === p.ReferencesRequest.method) {
       // https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_references
       let params = msg.params as p.ReferenceParams;
       let filePath = fileURLToPath(params.textDocument.uri);
-      let result: Location | null = utils.runAnalysisAfterSanityCheck(
+      let result: Location[] | null = utils.runAnalysisAfterSanityCheck(
```

Comment:
Thanks!

Reply:
Actually we had a few of these.
See this alternative way which I'm going to merge now:
https://github.com/rescript-lang/rescript-vscode/pull/232

Reply:
Merged --- perhaps rebase on master instead.

Reply:
I'm not sure if the types on #232 are the correct types.
I've also took look at protocol types in ts and couldn't find correct types.
I think we should rely on manual types based on protocl docs. for example in the docs the response type for references request is `Location[] | null` but there's not type in ts spec for that.